### DISCUSSION TITLE
Access Privacy Badger results from Selenium tests

### DISCUSSION THREAD

**Discussion by verzulli** [EFForg/privacybadger#2833] [Q&A] [answered]
Hi all,
I'd like to access PB results (the ones reported in popup.html) from within Selenium tests.

Basically, I'm trying to check "privacy compliance" of various websites (Public Administrations, mainly), and I want to automate such a check with Selenium.

Instead of reinventing the wheel adding my own logic to search for relevant tag, script, cookie, etc.... I thought to use Selenium with a Firefox profile that includes the Privacy Badger Extension, so to offload the detection activities to PB...

I've been successful in configuring the profile and, at the moment, when selenium request a web-page, it properly show PB results (see image below).

How can I get further, retrieving such info from within my code? Can PB "inject" its result somewhere so, for me, to retrieve it? As far as you know, is there a way, from Selenium, to get access to "extension data" ( `POPUP_DATA.origins` within `skin/popup.html` would be perfect!)?

Please note that I'm using Selenium with NodeJS
![scr](https://user-images.githubusercontent.com/6765895/155415326-113bd52f-2d31-4cbb-9c6a-a1fab176c5d1.png)

**Comment by ghostwords** [comment, marked as answer]
Hello! You could open `options.html`, get a [handle to the background page](https://github.com/EFForg/privacybadger/search?q=chrome.extension.getBackgroundPage%28%29) and thus gain access to the [various data structures](https://github.com/EFForg/privacybadger/blob/master/doc/DESIGN-AND-ROADMAP.md#data-structures).

Per-tab results are stored in the [`badger.tabData` object](https://github.com/EFForg/privacybadger/blob/077af32c15e462dd2e40e6e95c4f95cd16d26892/src/js/background.js#L243-L246). You can find an [example of querying tab data](https://github.com/EFForg/privacybadger/blob/077af32c15e462dd2e40e6e95c4f95cd16d26892/tests/selenium/service_workers_test.py#L13-L20) in Privacy Badger's Selenium-powered tests.

Note that you will need to [open `options.html` in a new window](https://github.com/EFForg/privacybadger/blob/077af32c15e462dd2e40e6e95c4f95cd16d26892/tests/selenium/service_workers_test.py#L46-L49), keeping the original website tab open to avoid Privacy Badger cleaning up `tabData`.

Let me know if this helps.

**Reply by verzulli** [reply]
IT WORKED!!!! :-)

First of all, *THANK YOU VERY MUCH* for answering!

I've had some troubles getting deeper with your suggestions (a bit "too-summarized" for my very-basic-level of understanding of selenium-related activities)... but in the end, I *succeded* :-)

Just for the benefit of other people, getting here, let me rephrase @ghostwords suggestions:

1. The PB extension is BROWSER-WIDE. So, if the extension is properly installed **AND** you load a web page in a TAB **AND** you open PB in **A DIFFERENT** TAB, **THEN** accessing the _tabData_ data-structure from within the PB TAB.... will give you the results **ALSO** of the other TABs (aka: including the one you loaded previously and that you want to check with PB);

So, in order to access PB datastructure you simply need to:

2. ensure to rely on a browser with the extension properly installed and configured (I guess the extension can be installed on-the-fly but I've been unable to do so. So I lasted in using a pre-configured Firefox profile, with the extension already installed);
3. open you URL, normally (...so that it will load in the default first TAB). Keep track of such a TAB, to switch back to it, later on;
4. inject a javascript to the browser, to open a new TAB;
 
in the new TAB

5. open the PB option page;
6. access the _tabData_ datastructure to retrieve your information, sending them back to the main application;
7. inject a javascript to "close" the TAB
8. tell Selenium to move back to the main TAB, by using the aforementioned handle (point 3. above);

loop as many times as you want...

I just published my application (under A-GPL v.3) - **[Privacy Web-Hunter](https://gitlab.com/verzulli/privacy-web-hunter/)**  - on GitLab. Feel free to give it a look :-)

The component interacting with PB is [pageChecker.js](https://gitlab.com/verzulli/privacy-web-hunter/-/blob/master/components/pageChecker.js) - see _process_ method

Thanks again for your support!
